src/lib/routers/history.js

```diff
@@ -0,0 +1,130 @@
+}
+
+function defaultParseURL(qsModule, location) {
+  return qsModule.parse(location.search.slice(1));
+}
+
+class BrowserHistory {
+  /**
+   * Initializes a new storage provider that will sync the search state in the URL
+   * using web API (window.location.pushState and onpopstate event).
+   * @param {object} $0 the options.
+   * @param {function(object):string} [$0.windowTitle] function that transforms a UI state
+   * into a title for the page. It takes one parameter: a syncable object (generated by the mapping
+   * provided to the URL sync). It should return a string that will be the title.
+   * @param {number} [$0.writeDelay = 400] time before a write is actually done.
+   * Prevent having too much entries in the history and thus make the back button more friendly.
+   * @param {function(qs, object):string} [$0.createURL] generates the full URL. If not provided,
+   * the storage adaptor will mapped all syncable keys to the query string of the URL. The first
+   * parameter is a utility object that has two methods: `stringify` that creates a query string
+   * from an object and `parse` that transforms a query string into an object.
+   * @param {function(qs): object} [$0.parseURL] parses an URL into an object. It should symetrical
+   * to `createURL`. It gets as an argument an object that contains two methods: `stringify` that
+   * creates a query string from an object and `parse` that transforms a query string into an object.
+   */
+  constructor({
+    windowTitle,
+    writeDelay = 400,
+    createURL = defaultCreateURL,
+    parseURL = defaultParseURL,
+  } = {}) {
+    this.windowTitle = windowTitle;
+    this.writeTimer = undefined;
+    this.writeDelay = writeDelay;
+    this._createURL = createURL;
+    this.parseURL = parseURL;
+  }
+
+  /**
+   * This method pushes a search state into the URL.
+   * @param {object} uiState a syncable UI state
+   * @return {undefined}
+   */
+  write(uiState) {
+    const url = this.createURL(uiState);
+    const title = this.windowTitle && this.windowTitle(uiState);
+
+    if (this.writeTimer) {
+      window.clearTimeout(this.writeTimer);
+    }
+
+    this.writeTimer = setTimeout(() => {
+      if (title) window.document.title = title;
+      window.history.pushState(uiState, title || '', url);
+      this.writeTimer = undefined;
+    }, this.writeDelay);
+  }
+
+  /**
+   * This methods read the URL and returns a syncable UI search state.
+   * @return {object} the equivalent to what is store in the URL as an object
+   */
+  read() {
+    return this.parseURL(qs, window.location);
+  }
+
+  /**
+   * This methods sets a callback on the `onpopstate` event of the history API
+   * of the current page. This way, the URL sync can keep track of the changes.
+   * @param {function(object)} cb the callback that will receive the latest uiState.
+   * It is called when the URL is updated.
+   * @returns {undefined}
+   */
+  onUpdate(cb) {
```

Comment:
I will go for a different name on this method. The `onUpdate` name intent a function that will be called multiple times. At the end the function is like an `addListener`, we call it to subscribe to udpate.

Reply:
I'm not sure to follow here. onUpdate is similar to onClick, the event being update and it happens on the storage. If we go for addListener, it implies that you can have multiple kinds of events, no?

Reply:
When I first look at the function I thought the function will be call on each update and then add a listener multiple times. What we want is to listen for the event and call the callback on each update. Right?

Reply:
It's a method used by InstantSearch.js and that is  also used for the update callback that you can set in widgets (like the search box). It is used to connect another part of the system to the update of the state of the router. We will 1-n functions plugged into that. It will be strictly limited to updates.

What do you think? Should I still modify it?

Reply:
IMO yes we should but at the end you know the system better than me. Take the path that you find the most understandable. I'm fine with it.

Reply:
Let's talk about that Monday 👍